src/measuro.hpp

```diff
@@ -2109,6 +2109,242 @@ namespace measuro
+     *
+     * Rendered metrics consist of
+     *
+     * Reference: https://github.com/prometheus/docs/blob/main/content/docs/instrumenting/exposition_formats.md#text-based-format
+     *
+     * @include prometheus_renderer_example.txt
+     *
+     * @remarks thread-hostile
+     */
+    class PrometheusRenderer : public Renderer
+    {
+    public:
+        /*!
+         * Constructor.
+         *
+         * @param[in]    destination    Output stream to which to write the rendered metrics. This stream object @b must persist beyond the life of the renderer
+         *
+         * @remarks thread-hostile
+         */
+        PrometheusRenderer(std::ostream & destination, std::function<std::int64_t ()> timestamp_getter,
+                const std::string & app_name = "") noexcept : m_destination(destination), m_timestamp_getter(timestamp_getter),
+                m_count(0), m_app_name(app_name), m_METRIC_NAME_PATTERN("[a-zA-Z_:][a-zA-Z0-9_:]*"),
+                m_METRIC_NAME_CHAR_PATTERN("[a-zA-Z0-9_:]"), m_METRIC_UNIT_CHAR_PATTERN("[a-zA-Z0-9]")
+        {
+        }
+
+        virtual ~PrometheusRenderer() noexcept
+        {
+        }
+
+        /*!
+         *
+         *
+         * @remarks thread-hostile
+         */
+        virtual void before() noexcept(false) override
+        {
+            m_count = 0;
+        }
+
+        /*!
+         *
+         *
+         * @remarks thread-hostile
+         */
+        virtual void after() noexcept(false) override
+        {
+            m_destination << '\n';
+            m_destination.flush();
+        }
+
+        /*!
+         * Renders the metric as an entry in a Prometheus metric set.
+         *
+         * @example prometheus_renderer_example.txt
+         *
+         * @param[in]    metric    The metric to render
+         *
+         * @remarks thread-hostile
+         */
+        void render(const std::shared_ptr<Metric> & metric) noexcept(false) override final
+        {
+            std::string metric_value;
+
+            switch(metric->kind())
+            {
+            case Metric::Kind::UINT:
+            case Metric::Kind::INT:
+            case Metric::Kind::FLOAT:
+            case Metric::Kind::RATE:
+            case Metric::Kind::SUM:
+                metric_value = std::string((*metric));
+                break;
+            case Metric::Kind::BOOL:
+                if (bool(*metric))
+                {
+                    metric_value = "1";
+                }
+                else
+                {
+                    metric_value = "0";
+                }
+                break;
+            default:
+                /*
+                 * Only numeric metrics are supported by this renderer.
+                 */
+                return;
+            }
+
+            std::string name_dirty{metric->name()};
+            std::string name_sanitised;
+            if (!sanitise_name(name_dirty, name_sanitised))
```

Comment:
Could name and unit sanitization happen during construction instead of render? If they're need to be able to change, they could have helper functions to update and re-sanitize. Would save some time in the renderer.

Reply:
No unfortunately not - the renderer is passed individual metrics during render; it has no knowledge of the metric registry in advance. Different registries may be used with the same renderer object, and the metrics within them may change too.